File: Baekjoon/유석환/Q1654.swift
let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let K = input[0]
let N = input[1]
var kArr = [Int]()

for _ in 0..<K {
    kArr.append(Int(readLine()!)!)
}

var start = 1
var end = kArr.max()!
var result = Int()

while start <= end {
    let mid = (start + end) / 2
    
    // 각 랜선을 mid만큼 자른 개수 배열
    let cntArr = kArr.map { $0 / mid }
    let sum = cntArr.reduce(0, +)
    
    // 총 개수가 N보다 같거나 크면 개수를 줄여서 size를 더 늘릴 수 있음
    if sum >= N {
        result = mid
        start = mid + 1
    } else { // 총 개수가 N보다 작으면 size를 줄이고 개수를 더 늘려야 함
        end = mid - 1
    }
}

print(result)

File: Baekjoon/노수일/Q1654.py
from sys import stdin

_input_numbers = [int(num) for num in input().split(' ')]
_k, _n = _input_numbers[0], _input_numbers[1]
_numbers = []

for i in range(0, _k):
    _numbers.append(int(stdin.readline()))

_start, _end = 1, max(_numbers)

while _start <= _end:
    _length = (_start + _end) // 2
    _number_of_line = 0

    for number in _numbers:
        _number_of_line = _number_of_line + number // _length

    if _number_of_line >= _n:
        _start = _length + 1
    else:
        _end = _length - 1

print(_end)
